src/design/introspection.stanza

```diff
@@ -173,9 +219,287 @@ public defn get-pads-from-port (p2p-map:Tuple<KeyValue<JITXObject, Tuple<JITXObj
+from the `cxt` context, then that object will not be found. This
+function could find `cxt.some` if passed `r = some`.
+
+@param r Ref to an object in `cxt` - Note that the path must be
+referenced from the `cxt`. If you have a prefix - then the
+instance will not be found.
+@param cxt Context from which we will query the instances. By default
+this value is `self`.
+@return Instance of a `pcb-component` in the `cxt` context. Note that
+`pcb-module` instances will be filtered out.
+<DOC>
+public defn get-component-by-ref (r:Ref, cxt:JITXObject = self) -> Maybe<JITXObject>:
+  inside pcb-module:
+    for obj in component-instances(cxt) first:
+      if ref(obj) == r:
+        One(obj)
+      else:
+        None()
+
+doc: \<DOC>
+Find an instance by `Ref` by searching the design hierarchy.
+
+This function must be called from within a `pcb-module` context.
+
+@param r Ref of the `Instance` object we wish to find. An instance can
+be either a `pcb-module` or `pcb-component`.
+@param cxt Parent scope to search in. By default this is `self`.
+
+@return If we find an instance by the pass Ref `r` in the `cxt` context, then
+we will return a `JITXObject` for that instance. Otherwise, None().
+<DOC>
+public defn find-instance-by-ref (r:Ref, cxt:JITXObject = self) -> Maybe<JITXObject> :
+  inside pcb-module:
+
+    var search-space = cxt
+    var last-comp:Maybe<JITXObject> = None()
+    val comps = Array<Maybe<JITXObject>>(ref-length(r))
+    for i in 1 through ref-length(r) do:
+      val b = base(r, i)
+      val comp? = for obj in instances(search-space) first:
+        if ref(obj) == b:
+          One(obj)
+        else:
+          None()
+
+      comps[i - 1] = comp?
+      match(comp?:One<JITXObject>):
+        search-space = value!(comp?)
+
+    ; Now I have a list of all of the elements in my ref from start
+    ;  to finish.
+    ; I want to filter this for the last non-None value and confirm
+    ;  that it is a component
+
+    val trail-comp? = for comp in in-reverse(comps) first:
+      if comp is-not None:
+        One(value!(comp))
+      else:
+        None()
+
+    trail-comp?
+
+doc: \<DOC>
+Get the net associated with a module or component port
+
+This function must be called from within a `pcb-module` context.
+
+Note that this function will fail to find a net if the passed `pt`
+port is not part of a component or module in the local context.
+For example, If you have:
+
+```stanza
+pcb-module wrapper:
+  port GND
+  inst C : some-component
+
+  ; `INT` is defined only in the `wrapper`
+  ;   context.
+  net INT (C.internal-port, ...)
+
+pcb-module top-level:
+
+  inst W : wrapper
+  net GND (W.GND)
+
+  ; This will work
+  val n1 = get-connected-net!(W.GND)
+
+  ; This will not work
+  val n2 = get-connected-net!(W.C.internal-port)
+```
+The value for `n1` will be successfully found as the `GND`
+net because it is defined within the `pcb-module` context.
+
+The `n2` invocation will throw an exception and not be able to find `INT`
```

Comment:
nit: The `n2` invocation will not be able to find `INT` as the net for `W.C.internal-port` and will throw an exception.